sycl/doc/extensions/experimental/sycl_ext_oneapi_virtual_mem.asciidoc

```diff
@@ -0,0 +1,330 @@
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback.  Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Backend support status
+
+The APIs in this extension may be used only on a device that has
+`aspect::ext_oneapi_virtual_mem`.  The application must check that the devices
+in the corresponding context have this aspect before using any of the APIs
+introduced in this extension.  If the application fails to do this, the
+implementation throws a synchronous exception with the
+`errc::feature_not_supported` error code.
+
+== Overview
+
+This extension adds the notion of "virtual memory ranges" to SYCL, introducing
+a way to map an address range onto multiple allocations of physical memory,
+allowing users to avoid expensive reallocations and potentially running out of
+device memory while relocating the corresponding memory.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_VIRTUAL_MEM` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Memory granularity
+
+Working with virtual address ranges and the underlying physical memory requires
+the user to align and adjust in accordance with a specified minimum granularity.
+In addition, devices can have a recommended granularity which may different from
+the minimum granularity and can be used instead of the minimum granularity.
+
+The interfaces for querying the these granularities are defined as:
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+size_t get_minimum_mem_granularity(size_t numBytes, const device &syclDevice, const context &syclContext);
+size_t get_minimum_mem_granularity(size_t numBytes, const queue &syclQueue);
+size_t get_minimum_mem_granularity(size_t numBytes, const physical_mem &syclPhysicalMem);
+
+size_t get_recommended_mem_granularity(size_t numBytes, const device &syclDevice, const context &syclContext);
+size_t get_recommended_mem_granularity(size_t numBytes, const queue &syclQueue);
+size_t get_recommended_mem_granularity(size_t numBytes, const physical_mem &syclPhysicalMem);
+
```

Comment:
what's the difference between minimum and recommended? In L0 and CUDA I think they are the same.

Reply:
[cuMemGetAllocationGranularity](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__VA.html#group__CUDA__VA_1g30ee906c2cf66a0347b3dfec3d7eb31a) takes a [`CUmemAllocationGranularity_flags`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1g3a202e4d32ae296db1af7efe75ce365d) flag that can either be `CU_MEM_ALLOC_GRANULARITY_MINIMUM` or `CU_MEM_ALLOC_GRANULARITY_RECOMMENDED`. I can't say if it actually returns any different values on existing devices.